File: Insertion Sort List/Insertion_Sort_List.py
# Runtime: 1324 ms, faster than 50.00% of Python3 online submissions for Insertion Sort List.
# Memory Usage: 14.6 MB, less than 100.00% of Python3 online submissions for Insertion Sort List.

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def insertionSortList(self, head: 'ListNode') -> 'ListNode':
        if head is None or head.next is None:
            return head

        preNode, curNode, nextNode = None, head, head
        while curNode:
            nextNode = nextNode.next

            leftNode, rightNode = None, head
            insert = False
            while rightNode != curNode:
                if rightNode.val <= curNode.val:
                    leftNode = rightNode
                    rightNode = rightNode.next
                else:
                    insert = True
                    curNode.next = rightNode

                    if leftNode:
                        leftNode.next = curNode
                    else:
                        head = curNode
                    
                    preNode.next = nextNode
                    break
            
            if not insert:
                preNode = curNode

            curNode = nextNode
        
        return head

File: Insertion Sort List/Insertion_Sort_List.cpp
// 链表的插入排序

// Runtime: 48 ms, faster than 75.93% of C++ online submissions for Insertion Sort List.
// Memory Usage: 9.5 MB, less than 100.00% of C++ online submissions for Insertion Sort List.

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution 
{
public:
    ListNode* insertionSortList(ListNode* head) 
    {
        // 如果只有一个元素或者没有元素的话直接返回即可，不需要排序
        if (head == 0 || head->next == 0)
            return head;

        for (ListNode *preNode = 0, *curNode = head, *nextNode = head; curNode; curNode = nextNode)
        {
            nextNode = nextNode->next;

            // 迭代寻找适合的插入位置
            ListNode *leftNode = 0, *rightNode = head;
            bool insert = false;
            while (rightNode != curNode)
            {
                if (rightNode->val <= curNode->val)
                {
                    leftNode = rightNode;
                    rightNode = rightNode->next;
                }
                else
                {
                    insert = true;
                    curNode->next = rightNode;
                    
                    if (leftNode)
                        leftNode->next = curNode;
                    else
                        head = curNode;

                    preNode->next = nextNode;
                    break;
                }
            }

            if (!insert)
                preNode = curNode;
        }
        return head;    
    }
};